pkg/descheduler/strategies/node_taint.go

```diff
@@ -0,0 +1,148 @@
+	podsEvicted := 0
+	for _, node := range nodes {
+		glog.V(1).Infof("Processing node: %#v\n", node.Name)
+		pods, err := podutil.ListEvictablePodsOnNode(client, node, evictLocalStoragePods)
+		if err != nil {
+			//no pods evicted as error encountered retrieving evictable Pods
+			return 0
+		}
+		totalPods := len(pods)
+		for i := 0; i < totalPods; i++ {
+			if maxPodsToEvict > 0 && nodePodCount[node]+1 > maxPodsToEvict {
+				break
+			}
+			if !checkPodsSatisfyTolerations(pods[i], node) {
+				success, err := evictions.EvictPod(client, pods[i], policyGroupVersion, dryRun)
+				if !success {
+					glog.Errorf("Error when evicting pod: %#v (%#v)\n", pods[i].Name, err)
+				} else {
+					nodePodCount[node]++
+					glog.V(1).Infof("Evicted pod: %#v (%#v)", pods[i].Name, err)
+				}
+			}
+		}
+		podsEvicted += nodePodCount[node]
+	}
+	return podsEvicted
+}
+
+// checkPodsSatisfyTolerations checks if the node's taints (NoSchedule) are still satisfied by pods' tolerations.
+func checkPodsSatisfyTolerations(pod *v1.Pod, node *v1.Node) bool {
+	tolerations := pod.Spec.Tolerations
+	taints := node.Spec.Taints
+	if len(taints) == 0 {
+		return true
+	}
+	noScheduleTaints := getNoScheduleTaints(taints)
+	allTolerated, _ := getMatchingTolerations(noScheduleTaints, tolerations)
+	if !allTolerated {
+		glog.V(2).Infof("Not all taints are tolerated after update for Pod %v on node %v", pod.Name, node.Name)
+		return false
+	}
+	return true
+}
+
+// getNoScheduleTaints return a slice of NoSchedule taints from the a slice of taints that it receives.
+func getNoScheduleTaints(taints []v1.Taint) []v1.Taint {
+
+	result := []v1.Taint{}
+	for i := range taints {
+		if taints[i].Effect == v1.TaintEffectNoSchedule {
+			result = append(result, taints[i])
+		}
+	}
+	return result
+
+}
+
+//toleratesTaint returns true if a toleration tolerates a taint, or false otherwise
+func toleratesTaint(toleration *v1.Toleration, taint *v1.Taint) bool {
+
+	if len(toleration.Effect) > 0 && toleration.Effect != taint.Effect {
+		return false
+	}
+
+	if len(toleration.Key) > 0 && toleration.Key != taint.Key {
+		return false
+	}
+
+	switch toleration.Operator {
+	// empty operator means Equal
+	case "", TolerationOpEqual:
+		return toleration.Value == taint.Value
+	case TolerationOpExists:
+		return true
+	default:
+		return false
+	}
+}
+
+// getMatchingTolerations returns true and list of Tolerations matching all Taints if all are tolerated, or false otherwise.
+func getMatchingTolerations(taints []v1.Taint, tolerations []v1.Toleration) (bool, []v1.Toleration) {
+	if len(taints) == 0 {
+		return true, []v1.Toleration{}
+	}
+	if len(tolerations) == 0 && len(taints) > 0 {
+		return false, []v1.Toleration{}
+	}
+	result := []v1.Toleration{}
+	for i := range taints {
+		tolerated := false
+		for j := range tolerations {
+			if toleratesTaint(&tolerations[j], &taints[i]) {
+				result = append(result, tolerations[j])
+				tolerated = true
+				break
```

Comment:
If we're not using the result anywhere, we could remove the whole `result` logic and change this section to simply `if !toleratesTaint(..) { return false }`